File: Simulator-CSharp/Components/CardReader.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Data;
using System.Linq;
using System.Text;
using System.Windows.Forms;

namespace ATM.Components
{
    public partial class CardReader : UserControl
    {
        public CardReader()
        {
            InitializeComponent();
        }

        private string number;
        public string Number { get { return number; } private set { } }
                
        public bool IsCardInserted { get { return !String.IsNullOrEmpty(number); } private set { } }

        public event EventHandler CardInserted;
        public void OnCardInserted(object sender, EventArgs e)
        {
            panelCard.Visible = true;
            panelReader.BackColor = SystemColors.ControlDarkDark;
            if (CardInserted != null) CardInserted(this, e);
        }

        public void EjectCard()
        {
            Console.WriteLine("Expulsando tarjeta");
            OnCardEjected(this, EventArgs.Empty);
        }

        public event EventHandler CardEjected;
        public void OnCardEjected(object sender, EventArgs e)
        {
            number = null;
            panelCard.Visible = false;
            panelReader.BackColor = SystemColors.ControlDark;
            if (CardEjected != null) CardEjected(this, e);
        }

        private void CardReader_DragDrop(object sender, DragEventArgs e)
        {
            if (!IsCardInserted)
            {
                number = e.Data.GetData(DataFormats.Text).ToString();
                OnCardInserted(this, EventArgs.Empty);
            }
        }

        private void CardReader_DragEnter(object sender, DragEventArgs e)
        {
            if (e.Data.GetDataPresent(DataFormats.Text) && !IsCardInserted)
            {
                e.Effect = DragDropEffects.Copy;
            }
            else {
                e.Effect = DragDropEffects.None;
            }
        }

        private void panelCard_Click(object sender, EventArgs e)
        {
            OnCardEjected(this, EventArgs.Empty);
        }
    }
}

File: Simulator-VB/Components/CardReader.vb
Public Class CardReader

    Private _Number As String
    Public ReadOnly Property Number As String
        Get
            Return _Number
        End Get
    End Property

    Public ReadOnly Property IsCardInserted As Boolean
        Get
            Return Not String.IsNullOrEmpty(_Number)
        End Get
    End Property

    Public Event CardInserted As EventHandler
    Protected Sub OnCardInserted(sender As Object, e As EventArgs)
        Me.PanelCard.Visible = True
        Me.PanelReader.BackColor = SystemColors.ControlDarkDark
        RaiseEvent CardInserted(Me, e)
    End Sub

    Public Sub EjectCard()
        Console.WriteLine("Expulsando tarjeta")
        OnCardEjected(Me, EventArgs.Empty)
    End Sub

    Public Event CardEjected As EventHandler
    Protected Sub OnCardEjected(sender As Object, e As EventArgs)
        _Number = Nothing
        Me.PanelCard.Visible = False
        Me.PanelReader.BackColor = SystemColors.ControlDark
        RaiseEvent CardEjected(Me, e)
    End Sub

    Private Sub CardReader_DragDrop(sender As Object, e As DragEventArgs) Handles Me.DragDrop
        If Not IsCardInserted Then
            _Number = e.Data.GetData(DataFormats.Text).ToString()
            Console.WriteLine("Leyendo tarjeta: " & _Number)
            OnCardInserted(Me, EventArgs.Empty)
        End If
    End Sub

    Private Sub CardReader_DragEnter(sender As Object, e As DragEventArgs) Handles Me.DragEnter
        If e.Data.GetDataPresent(DataFormats.Text) And Not IsCardInserted Then
            e.Effect = DragDropEffects.Copy
        Else
            e.Effect = DragDropEffects.None
        End If
    End Sub

    Private Sub PanelCard_Click(sender As Object, e As EventArgs) Handles PanelCard.Click
        OnCardEjected(Me, EventArgs.Empty)
    End Sub

End Class
